File: moviebuff-backend/src/main/java/com/moviebuff/moviebuff_backend/service/theater/ShowService.java
package com.moviebuff.moviebuff_backend.service.theater;

import com.moviebuff.moviebuff_backend.dto.request.ShowRequest;
import com.moviebuff.moviebuff_backend.dto.response.ShowResponse;
import com.moviebuff.moviebuff_backend.model.show.Show;
import com.moviebuff.moviebuff_backend.repository.interfaces.theater.IShowRepository;
import com.moviebuff.moviebuff_backend.repository.impl.theater.ShowRepositoryImpl;
import com.moviebuff.moviebuff_backend.exception.ResourceNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ShowService implements IShowService {

    @Autowired
    private IShowRepository showRepository;

    @Autowired 
    private ShowRepositoryImpl showRepositoryImpl;
    
    @Autowired
    private ShowMapper showMapper;

    @Override
    @Transactional
    public ShowResponse createShow(ShowRequest request) {
        validateShowTimings(request);
        Show show = showMapper.mapRequestToEntity(request);
        show = showRepository.save(show);
        return showMapper.mapEntityToResponse(show);
    }

    @Override
    @Transactional 
    public ShowResponse updateShow(String id, ShowRequest request) {
        Show show = showRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Show not found with id: " + id));
            
        // Only allow updates if show hasn't started
        if (show.getShowTime().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Cannot update show that has already started");
        }
        
        validateShowTimings(request);
        showMapper.updateEntityFromRequest(show, request);
        show = showRepository.save(show);
        return showMapper.mapEntityToResponse(show);
    }

    @Override
    @Transactional
    public void deleteShow(String id) {
        Show show = showRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Show not found with id: " + id));
            
        // Only allow deletion if show hasn't started
        if (show.getShowTime().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Cannot delete show that has already started");
        }
        
        showRepository.deleteById(id);
    }

    @Override
    public ShowResponse getShow(String id) {
        return showRepository.findById(id)
            .map(showMapper::mapEntityToResponse)
            .orElseThrow(() -> new ResourceNotFoundException("Show not found with id: " + id));
    }

    @Override
    public List<ShowResponse> getShowsByTheater(String theaterId) {
        return showRepository.findByTheaterId(theaterId).stream()
            .map(showMapper::mapEntityToResponse)
            .collect(Collectors.toList());
    }

    @Override
    public List<ShowResponse> getShowsByMovie(String movieId) {
        return showRepository.findUpcomingShowsByMovie(movieId, LocalDateTime.now()).stream()
            .map(showMapper::mapEntityToResponse)
            .collect(Collectors.toList());
    }

    @Override 
    public List<ShowResponse> getShowsByTheaterAndScreen(String theaterId, int screenNumber, 
            LocalDateTime startTime, LocalDateTime endTime) {
        return showRepository.findShowsByTheaterAndScreen(theaterId, screenNumber, startTime, endTime).stream()
            .map(showMapper::mapEntityToResponse)
            .collect(Collectors.toList());
    }

    @Override
    @Transactional
    public void updateShowStatus(String showId, String status) {
        // First verify the show exists
        Show show = showRepository.findById(showId)
            .orElseThrow(() -> new ResourceNotFoundException("Show not found with id: " + showId));
            
        try {
            // Convert string status to enum and validate
            Show.ShowStatus newStatus = Show.ShowStatus.valueOf(status.toUpperCase());
            // Update the show's status and save it
            show.setStatus(newStatus);
            showRepository.save(show);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid show status: " + status);
        }
    }

    @Override
    @Transactional
    public void updateSeatAvailability(String showId, List<String> seatIds, boolean available) {
        Show show = showRepository.findById(showId)
            .orElseThrow(() -> new ResourceNotFoundException("Show not found with id: " + showId));
            
        // Validate show timing
        if (show.getShowTime().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Cannot update seats for a show that has already started");
        }
        
        showRepositoryImpl.updateSeatAvailability(showId, seatIds, available);
    }

    private void validateShowTimings(ShowRequest request) {
        // Validate show time is in the future
        if (request.getShowTime().isBefore(LocalDateTime.now())) {
            throw new IllegalArgumentException("Show time must be in the future");
        }
        
        // Calculate show end time based on movie duration
        LocalDateTime showEndTime = request.getShowTime().plusMinutes(request.getDuration());
        
        // Add buffer time between shows (e.g., 30 minutes)
        LocalDateTime bufferStartTime = request.getShowTime().minusMinutes(30);
        LocalDateTime bufferEndTime = showEndTime.plusMinutes(30);
        
        List<Show> conflictingShows = showRepositoryImpl.findConflictingShows(
            request.getTheaterId(), 
            request.getScreenNumber(),
            bufferStartTime,
            bufferEndTime
        );
        
        if (!conflictingShows.isEmpty()) {
            throw new RuntimeException("Show timing conflicts with existing shows (including buffer time)");
        }
    }
}

File: frontend/src/services/showService.js
// src/services/showService.js - Updated with endTime handling
import axiosInstance from './axiosConfig';

const BASE_URL = '/shows';

const showService = {
  // Create a new show
  createShow: async (showData) => {
    try {
      console.log('----------------------------------------------------------------');
      console.log("showData: " + JSON.stringify(showData));
      
      console.log('----------------------------------------------------------------');

      
      const response = await axiosInstance.post(BASE_URL, showData);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to create show';
    }
  },

  // Update an existing show
  updateShow: async (id, showData) => {
    try {
      const response = await axiosInstance.put(`${BASE_URL}/${id}`, showData);
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to update show ${id}`;
    }
  },

  // Delete a show
  deleteShow: async (id) => {
    try {
      await axiosInstance.delete(`${BASE_URL}/${id}`);
      return id;
    } catch (error) {
      throw error.response?.data || `Failed to delete show ${id}`;
    }
  },

  // Get a show by ID
  getShow: async (id) => {
    try {
      const response = await axiosInstance.get(`${BASE_URL}/${id}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch show with ID ${id}`;
    }
  },

  // Get shows by theater
  getShowsByTheater: async (theaterId) => {
    try {
      const response = await axiosInstance.get(`${BASE_URL}/theater/${theaterId}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch shows for theater ${theaterId}`;
    }
  },

  // Get shows by movie
  getShowsByMovie: async (movieId) => {
    try {
      const response = await axiosInstance.get(`${BASE_URL}/movie/${movieId}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch shows for movie ${movieId}`;
    }
  },

  // Get shows by theater and screen for a date range
  getShowsByTheaterAndScreen: async (theaterId, screenNumber, startTime, endTime) => {
    try {
      const params = { startTime, endTime };
      const response = await axiosInstance.get(
        `${BASE_URL}/theater/${theaterId}/screen/${screenNumber}`,
        { params }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch shows for theater ${theaterId} screen ${screenNumber}`;
    }
  },

  // Get shows by date and optional city
  getShowsByDate: async (date, city) => {
    try {
      let params = { date };
      if (city) params.city = city;
      
      const response = await axiosInstance.get(`${BASE_URL}/date`, { params });
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch shows for date ${date}`;
    }
  },

  // Get shows by movie and city
  getShowsByMovieAndCity: async (movieId, city) => {
    try {
      const response = await axiosInstance.get(`${BASE_URL}/movie/${movieId}/city/${city}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch shows for movie ${movieId} in city ${city}`;
    }
  },

  // Update show status
  updateShowStatus: async (showId, status) => {
    try {
      const response = await axiosInstance.put(
        `${BASE_URL}/${showId}/status`, 
        null, 
        { params: { status } }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to update status for show ${showId}`;
    }
  },

  // Update seat availability
  updateSeatAvailability: async (showId, seatIds, available) => {
    try {
      const response = await axiosInstance.put(
        `${BASE_URL}/${showId}/seats`, 
        seatIds, 
        { params: { available } }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to update seat availability for show ${showId}`;
    }
  },

  // Get seat availability for a show
  getSeatAvailability: async (showId) => {
    try {
      const response = await axiosInstance.get(`${BASE_URL}/${showId}/seat-availability`);
      return response.data;
    } catch (error) {
      throw error.response?.data || `Failed to fetch seat availability for show ${showId}`;
    }
  },

  // Get show analytics
  getShowAnalytics: async (startDate, endDate, movieId, theaterId) => {
    try {
      let params = {};
      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;
      if (movieId) params.movieId = movieId;
      if (theaterId) params.theaterId = theaterId;
      
      const response = await axiosInstance.get(`${BASE_URL}/analytics`, { params });
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to fetch show analytics';
    }
  },

  // Get trending shows
  getTrendingShows: async (city, limit = 10) => {
    try {
      let params = { limit };
      if (city) params.city = city;
      
      const response = await axiosInstance.get(`${BASE_URL}/trending`, { params });
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to fetch trending shows';
    }
  }
};

export default showService;